### PR TITLE
GleanPlumb Messaging Part I

### PR DESCRIPTION
**Pull request by jhugman**
Transfer FML from design doc into repo

/cc @Amejia481

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by Amejia481** [comment]
Thanks!

File: app/src/main/java/org/mozilla/fenix/gleanplum/MessagesManager.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import kotlin.Comparator
+
+class MessagesManager(
+    private val context: Context,
+    private val storage: MessageStorage,
+    private val messagingFeature: FeatureHolder<Messaging>
+) : MessageController {
+
+    private val availableMessages: SortedSet<Message> = sortedSetOf(
+        Comparator { message1, message2 -> message2.style.priority.compareTo(message1.style.priority) }
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/gleanplum/MessagesManager.kt:17]: @jhugman apart from the priority, is there anything else that will help us to identify which messages should be shown first? I was thinking about triggers but not sure if this is out of scope for now?

File: app/src/main/java/org/mozilla/fenix/gleanplum/MessagesManager.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    override fun onMessagePressed(message: Message) {
+        // TODO: Report telemetry event
+        storage.updateMetadata(
+            message.metadata.copy(
+                pressed = true
+            )
+        )
+        availableMessages.remove(message)
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/gleanplum/MessagesManager.kt:31]: We will need to also perform the action here

File: nimbus.fml.yaml [hunk 3]
```diff
@@ -68,8 +68,220 @@ features:
...
+            A URL of a page or a deeplink.
+            This may have substitution variables in.
+          # This should never be defaulted.
+          default: empty_string
+        text:
+          type: Text
+          description: "The message text displayed to the user"
+          # This should never be defaulted.
+          default: empty_string
+        button-label:
```
Comment (Amejia481) [inline nimbus.fml.yaml:224]: @jhugman  Could you give me a hand :), I'm having some trouble trying to get the `button-label` into the `MessageData`class in Kotlin.
```
> Task :app:nimbusFeaturesDebug FAILED
Nimbus FML generating Kotlin
manifest  nimbus.fml.yaml
class      org.mozilla.fenix.nimbus.FxNimbus
channel    developer
R.class    org.mozilla.fenix.R
Error: Validation Error: Mismatch between type BundleText("") and default null

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:nimbusFeaturesDebug'.
> Process 'command '/app/build/bin/nimbus/nimbus-fml'' finished with non-zero exit value 1

```

File: app/src/main/java/org/mozilla/fenix/gleanplum/MessagesManager.kt [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    private val availableMessages: SortedSet<Message> = sortedSetOf(
+        Comparator { message1, message2 -> message2.style.priority.compareTo(message1.style.priority) }
+    )
+
+    fun areMessagesAvailable(): Boolean {
+        return availableMessages.isNotEmpty()
+    }
+
+    override fun getNextMessage(): Message? {
+        return availableMessages.first()
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/gleanplum/MessagesManager.kt:29]: I just noticed this was a draft, but typed too much to delete :(

 * run the JEXL triggers here.
 * [handle control messages and exposure events][0]

[0]:  https://docs.google.com/document/d/1zOdYpXkunWm-0MeUjTjuS_CmbfpJrsAta0vxgzEzy14/edit

File: app/src/main/java/org/mozilla/fenix/gleanplum/MessagesManager.kt [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        if (newMetadata.displayCount < message.data.maxDisplayCount) {
+            availableMessages.add(newMessage)
+        }
+    }
+
+    override fun initialize() {
+        val nimbusTriggers = messagingFeature.value().triggers
+        val nimbusStyles = messagingFeature.value().styles
+        val nimbusMessages = messagingFeature.value().messages
+        val defaultStyle = StyleData(context)
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/gleanplum/MessagesManager.kt:73]: I love this.

File: nimbus.fml.yaml (hunk 3 above)
Comment (jhugman) [reply nimbus.fml.yaml:224]: This looks like one of the bugs in FML that is fixed by https://github.com/mozilla/application-services/pull/4867 .

The current work around is to use `String` instead of `Text`.

**Review by jhugman — COMMENTED** [review]
So far, so good! Good job. 🎉

File: nimbus.fml.yaml (hunk 3 above)
Comment (Amejia481) [reply nimbus.fml.yaml:224]: Thanks for the suggestion, I gave it a try, but unfortunately I still getting a similar issue,  I updated the `type` to `String` and the `default` to `"default"`:

image


```
> Task :app:nimbusExperimenter
Nimbus FML generating JSON
manifest   /fenix/nimbus.fml.yaml
outputFile /fenix/.experimenter.json

> Task :app:nimbusFeaturesDebug FAILED
Nimbus FML generating Kotlin
manifest   /fenix/nimbus.fml.yaml
class      org.mozilla.fenix.nimbus.FxNimbus
channel    developer
R.class    org.mozilla.fenix.R
Error: Validation Error: Mismatch between type String and default null

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:nimbusFeaturesDebug'.
> Process 'command '/fenix/app/build/bin/nimbus/nimbus-fml'' finished with non-zero exit value 1


```

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -6,6 +6,9 @@
     <string name="app_name_private_5">Private %s</string>
     <!-- App name for private browsing mode. The first parameter is the name of the app defined in app_name (for example: Fenix)-->
     <string name="app_name_private_4">%s (Private)</string>
+    <string name="gdpr_has_you_covered_firefox_has_gdpr_covered">%s (Private)</string>
+    <string name="sharing_your_phone_autoclosing_private_tabs_is_for_you">%s (Private)</string>
+    <string name="love_firefox_fill_in_our_survey">%s (Private)</string>
```
Comment (Amejia481) [inline app/src/main/res/values/strings.xml:11]: @jhugman To be able of compiling, I had to add these mock strings. I was seeing those in the compiled code but I'm not quite sure from where they were fetched or how are we to add those strings.

File: nimbus.fml.yaml [hunk 7]
```diff
@@ -68,8 +68,220 @@ features:
...
+            button-text-color: white
+            priority: 60
+
+      actions:
+        type: Map<String, String>
+        description: A growable map of action URLs.
+        default:
+          OPEN_SETTINGS: firefox://settings
+          OPEN_SYNC_SETTINGS: firefox://settings/sync
+          OPEN_POCKET_SETTINGS: https:///getpocket.com/settings?fxa={fxa-token}
```
Comment (Amejia481) [inline nimbus.fml.yaml:131]: @jhugman just to make you aware the scheme for deeplink changes depending of on which build type (nightly,beta, and release)
nightly -> fenix-nightly://
beta -> fenix-beta://
release -> fenix://

See https://github.com/mozilla-mobile/fenix/blob/main/app/build.gradle#L113-L137

File: app/src/main/java/org/mozilla/fenix/gleanplum/DefaultMessageController.kt [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+ */
+class DefaultMessageController(
+    private val messageManager: MessagesManager,
+    private val homeActivity: HomeActivity
+) : MessageController {
+
+    override fun onMessagePressed(message: Message) {
+        // TODO: report telemetry event
+        messageManager.onMessagePressed(message)
+        handleAction(message.data.action)
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/gleanplum/DefaultMessageController.kt:24]: @jhugman when I try to access `message.data.action` do I get the deeplink uri of key for the action? E.G
https://github.com/mozilla-mobile/fenix/pull/24048/files#diff-680637c1a9cdbef47ee7a3333ef917b741c19c3266663c0c760c0cbbc69b1928R129-R131

Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/gleanplum/DefaultMessageController.kt:24]: See for more context 
https://github.com/mozilla-mobile/fenix/pull/24048/files#diff-337feac155df599668495c000b11628dbdf2d734512c3d6aece265a59a7a3a9bR39-R49

Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/gleanplum/DefaultMessageController.kt:24]: What will happen for a web URL when accessing `message.data.action`?

**Comment by Amejia481** [comment]
Closed in favor of https://github.com/mozilla-mobile/fenix/pull/24227 and https://github.com/mozilla-mobile/fenix/pull/24225
